tests/tests/test_state_scripts.py

```diff
@@ -632,6 +764,139 @@ class TestStateScripts(MenderTesting):
+                if script in test_set.get("RebootOnceScripts", []):
+                    fd.write(script_reboot_once)
+                elif script in test_set.get("StopSecondPartition", []):
+                    fd.write(script_stop_mender_on_entry)
+                elif script in test_set.get("ErrorScripts", []):
+                    fd.write(script_error_content)
+                else:
+                    fd.write(script_content)
+
+        # Now create the artifact, and make the deployment.
+        device_id = Helpers.ip_to_device_id_map([client])[client]
+        deployment_id = common_update_procedure(
+            install_image=new_rootfs,
+            broken_image=broken_image,
+            verify_status=True,
+            devices=[device_id],
+            scripts=[artifact_script_dir])[0]
+
+        try:
+
+            orig_part = Helpers.get_active_partition()
+
+            token = Helpers.place_reboot_token()
+
+            # handle case where the client has not finished the update
+            # path on the committed partition, but new partition is installed,
+            # thus we will not get a valid entrypoint into the uncommitted parition(reboot_leave)
+            # and the client will thus reboot straight after starting, and u-boot will
+            # fall back to the committed partition
+            if test_set.get("DoubleReboot", False):
+                token.verify_device_double_reboot()
+            else:
+                token.verify_reboot_performed(sleeptime=3, ntimes=3)
+
+            # if scripts are killed on a script that is run on the second partition,
+            # we need to wait for a second reboot
+            if test_set.get("DeviceDieSecondPartition", False) and len(
+                    test_set.get("StopSecondPartition", [])) > 0:
+
+                # Give it some time to settle
+                timeout = time.time() + 60
+                while timeout > time.time():
+                    logger.info("sleeping")
+                    time.sleep(3) # not in a hurry here, the client should be stopped
+                    output = run("journalctl -xn")
+                    logger.info(output)
+                    if "mender.service has finished shutting down" not in output:
+                        continue
+                    else:
+                        break
+                if time.time() > timeout:
+                    pytest.fail("could not restart mender")
+                token = Helpers.place_reboot_token()
+                run("systemctl start mender")
+                token.verify_reboot_performed(ntimes=3)
+
+            # settle down
+            timeout = time.time() + 60
+            while timeout > time.time():
+                logger.info("sleeping")
+                time.sleep(1)
+                if exists("/data/test_state_scripts.log"):
+                    break
+                else:
+                    pytest.fail("never found log files")
```

Comment:
What are you waiting for here? Given our greatly varying droplet performance, we have to use conditions when we wait instead of simply elapsed time, which is unreliable.